src/Tickets/Commerce/Gateways/PayPal/Client.php

```diff
@@ -494,5 +531,292 @@ public function refund_payment( $capture_id ) {
+		$query_args = [];
+		$args       = [
+			'headers' => [
+				'PayPal-Partner-Attribution-Id' => Gateway::ATTRIBUTION_ID,
+				'Prefer'                        => 'return=representation',
+			],
+		];
+
+		if ( ! wp_http_validate_url( $payment ) ) {
+			$payment = urlencode( $payment );
+			$url     = '/v2/payments/authorizations/{payment_id}';
+			$url     = str_replace( '{payment_id}', $payment, $url );
+		} else {
+			$url = $payment;
+		}
+
+		$response = $this->get( $url, $query_args, $args );
+
+		return $response;
+	}
+
+	/**
+	 * Verify the identity of the Webhook request, to avoid any security problems.
+	 *
+	 * @link  https://developer.paypal.com/docs/api/webhooks/v1/#verify-webhook-signature_post
+	 *
+	 * @since TBD
+	 *
+	 * @param string $webhook_id Which webhook id we have currently stored on the database.
+	 * @param array  $event      The Event received by the endpoint from PayPal.
+	 * @param array  $headers    Headers from the PayPal request that we use to verify the signature.
+	 *
+	 * @return bool
+	 */
+	public function verify_webhook_signature( $webhook_id, $event, $headers ) {
+		$query_args = [];
+		$body       = [
+			'transmission_id'   => Arr::get( $headers, 'transmission_id' ),
+			'transmission_time' => Arr::get( $headers, 'transmission_time' ),
+			'transmission_sig'  => Arr::get( $headers, 'transmission_sig' ),
+			'cert_url'          => Arr::get( $headers, 'cert_url' ),
+			'auth_algo'         => Arr::get( $headers, 'auth_algo' ),
+			'webhook_id'        => $webhook_id,
+			'webhook_event'     => $event,
+		];
+		$args       = [
+			'headers' => [
+				'PayPal-Partner-Attribution-Id' => Gateway::ATTRIBUTION_ID,
+			],
+			'body'    => $body,
+		];
+
+		$url      = 'v1/notifications/verify-webhook-signature';
+		$response = $this->post( $url, $query_args, $args );
+
+		return 'SUCCESS' === Arr::get( $response, 'verification_status', false );
+	}
+
+
+	/**
+	 * Get the list of webhooks.
+	 *
+	 * @see   https://developer.paypal.com/docs/api/webhooks/v1/#webhooks_list
+	 * @since TBD
+	 *
+	 * @return array[] The list of PayPal webhooks.
+	 */
+	public function list_webhooks() {
+		$query_args = [];
+		$body       = [];
+		$args       = [
+			'headers' => [
+				'PayPal-Partner-Attribution-Id' => Gateway::ATTRIBUTION_ID,
+				'Prefer'                        => 'return=representation',
+			],
+			'body'    => $body,
+		];
+
+		$url      = '/v1/notifications/webhooks';
+		$response = $this->get( $url, $query_args, $args );
+
+		if ( empty( $response['webhooks'] ) ) {
+			return [];
+		}
+
+		return $response['webhooks'];
+	}
+
+
+	/**
+	 * Get the webhook data from a specific webhook ID.
+	 *
+	 * @see   https://developer.paypal.com/docs/api/webhooks/v1/#webhooks_get
+	 * @since TBD
+	 *
+	 * @param string $webhook_id The webhook ID.
+	 *
+	 * @return array|null The PayPal webhook data.
+	 */
+	public function get_webhook( $webhook_id ) {
+		$query_args = [];
+		$body       = [];
+		$args       = [
```

Comment:
Do we maybe want to add these as optional parameters and use `wp_parse_args` with the defaults? Not sure if we need the flexibility to do it, now or in the future. As a dev I always appreciate when the code is flexible and I can easily use things for particular edge scenarios without modifying too much code.

Reply:
Not specifically for the connection to PayPal. These are super "strict" methods, mostly for internal usage.